Behaviour of doFinally

### Expected behavior
```java
@Test
public void errorBeforeFinallyTest() {
    System.err.println("> error before finally");
    Mono.just(true)
        .map(this::throwError)
        .doFinally(this::doFinalThing)
        .subscribe();
}

@Test
public void errorAfterFinallyTest() {
    System.err.println("> error after finally");
    Mono.just(true)
        .doFinally(this::doFinalThing)
        .map(this::throwError)
        .subscribe();
}

private void doFinalThing(SignalType any) {
    System.err.println("do finally called");
}

private Boolean throwError(Boolean x) {
    System.err.println("throw exception");
    throw new IllegalStateException();
}
```

According to the javadoc shouldn't be both tests have the same log output?

### Actual behavior

```
> error after finally
throw exception
do finally called


> error before finally
throw exception
```

### Steps to reproduce

See code above

### Reactor Core version

3.1.1

### JVM version (e.g. `java -version`)

java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)

### OS version (e.g. `uname -a`)
MINGW64_NT-10.0

Hi, I'm commenting in that because I think there is still an issue on the doFinally.

```java
    @Test
    public void withoutConsumerInSubscribe() {
        System.err.println("> subscribe without consumer: map_doOnError_doFinally");
        Mono.just(true)
                .map(this::throwError)
                .doOnError(e -> System.err.println("doOnError"))
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .subscribe();
        System.err.println("> subscribe without consumer: doFinally_map_doOnError");
        Mono.just(true)
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .map(this::throwError)
                .doOnError(e -> System.err.println("doOnError"))
                .subscribe();
        System.err.println("> subscribe without consumer:  map_doFinally_doOnError");
        Mono.just(true)
                .map(this::throwError)
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .doOnError(e -> System.err.println("doOnError"))
                .subscribe();
    }

    @Test
    public void withConsumerInSubscribe() {
        System.err.println("> subscribe with consumer: map_doOnError_doFinally");
        Mono.just(true)
                .map(this::throwError)
                .doOnError(e -> System.err.println("doOnError"))
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .subscribe(v -> { }, e -> { });
        System.err.println("> subscribe with consumer: doFinally_map_doOnError");
        Mono.just(true)
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .map(this::throwError)
                .doOnError(e -> System.err.println("doOnError"))
                .subscribe(v -> { }, e -> { });
        System.err.println("> subscribe with consumer: map_doFinally_doOnError");
        Mono.just(true)
                .map(this::throwError)
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .doOnError(e -> System.err.println("doOnError"))
                .subscribe(v -> { }, e -> { });
    }

    @Test
    public void whithoutDoOnError() {
        System.err.println("> whithoutDoOnError");
        Mono.just(true)
                .map(this::throwError)
                .doFinally(any -> System.err.println("doFinally " + any.toString()))
                .subscribe();
    }

    private Boolean throwError(Boolean x) {
        throw new IllegalStateException();
    }
```
the produce
```
> subscribe without consumer: map_doOnError_doFinally
doOnError
> subscribe without consumer: doFinally_map_doOnError
doFinally cancel
doOnError
> subscribe without consumer:  map_doFinally_doOnError
doOnError
doFinally onError
> subscribe with consumer: map_doOnError_doFinally
doOnError
doFinally onError
> subscribe with consumer: doFinally_map_doOnError
doFinally cancel
doOnError
> subscribe with consumer: map_doFinally_doOnError
doOnError
doFinally onError

> whithoutDoOnError
TestFinally > whithoutDoOnError FAILED
    reactor.core.Exceptions$ErrorCallbackNotImplemented
        Caused by: java.lang.IllegalStateException at TestFinally.java:69

java.lang.IllegalStateException
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.lang.IllegalStateException
Caused by: java.lang.IllegalStateException
	at TestFinally.throwError(TestFinally.java:69)
[...]
```

As you can see, when using `.subscribe()` instead of `.subscribe(consumer, errorConsumer)` the doFinally is not executed if it is placed after the `doOnError`.

Also, I upgraded from 3.0.7 to 3.1.2 and its not possible to write the code from `whithoutDoOnError ` with the `subscribe()` without parameters.


### Reactor Core version

3.1.2

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxDoFinally.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxDoFinally.java
@@ -125,8 +125,12 @@ final class FluxDoFinally<T> extends FluxOperator<T, T> {
 
 		@Override
 		public void onError(Throwable t) {
-			actual.onError(t);
-			runFinally(SignalType.ON_ERROR);
+			try {
+				actual.onError(t);
+			}
+			finally {
+				runFinally(SignalType.ON_ERROR);
+			}
 		}
 
 		@Override

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxDoFinallyTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxDoFinallyTest.java
@@ -16,6 +16,8 @@
 
 package reactor.core.publisher;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Queue;
 import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.function.Consumer;
@@ -392,4 +394,100 @@ public class FluxDoFinallyTest implements Consumer<SignalType> {
 	}
 
 	//TODO test multiple subscriptions?
+
+	@Test
+	//see https://github.com/reactor/reactor-core/issues/951
+	public void gh951_withoutConsumerInSubscribe() {
+		List<String> events = new ArrayList<>();
+		Mono.just(true)
+		    .map(this::throwError)
+		    .doOnError(e -> events.add("doOnError"))
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .subscribe();
+
+		Assertions.assertThat(events)
+		          .as("subscribe without consumer: map_doOnError_doFinally")
+		          .containsExactly("doOnError", "doFinally onError");
+
+		events.clear();
+		Mono.just(true)
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .map(this::throwError)
+		    .doOnError(e -> events.add("doOnError"))
+		    .subscribe();
+
+		Assertions.assertThat(events)
+		          .as("subscribe without consumer: doFinally_map_doOnError")
+		          .containsExactly("doFinally cancel", "doOnError");
+
+		events.clear();
+		Mono.just(true)
+		    .map(this::throwError)
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .doOnError(e -> events.add("doOnError"))
+		    .subscribe();
+
+		Assertions.assertThat(events)
+		          .as("subscribe without consumer:  map_doFinally_doOnError")
+		          .containsExactly("doOnError", "doFinally onError");
+	}
+
+	@Test
+	//see https://github.com/reactor/reactor-core/issues/951
+	public void gh951_withConsumerInSubscribe() {
+		List<String> events = new ArrayList<>();
+
+		Mono.just(true)
+		    .map(this::throwError)
+		    .doOnError(e -> events.add("doOnError"))
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .subscribe(v -> { }, e -> { });
+
+		Assertions.assertThat(events)
+		          .as("subscribe with consumer: map_doOnError_doFinally")
+		          .containsExactly("doOnError", "doFinally onError");
+
+		events.clear();
+		Mono.just(true)
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .map(this::throwError)
+		    .doOnError(e -> events.add("doOnError"))
+		    .subscribe(v -> { }, e -> { });
+
+		Assertions.assertThat(events)
+		          .as("subscribe with consumer: doFinally_map_doOnError")
+		          .containsExactly("doFinally cancel", "doOnError");
+
+		events.clear();
+		Mono.just(true)
+		    .map(this::throwError)
+		    .doFinally(any -> events.add("doFinally " + any.toString()))
+		    .doOnError(e -> events.add("doOnError"))
+		    .subscribe(v -> { }, e -> { });
+
+		Assertions.assertThat(events)
+		          .as("subscribe with consumer: map_doFinally_doOnError")
+		          .containsExactly("doOnError", "doFinally onError");
+	}
+
+	@Test
+	//see https://github.com/reactor/reactor-core/issues/951
+	public void gh951_whithoutDoOnError() {
+		List<String> events = new ArrayList<>();
+
+		Assertions.assertThatExceptionOfType(UnsupportedOperationException.class)
+		          .isThrownBy(Mono.just(true)
+		                          .map(this::throwError)
+		                          .doFinally(any -> events.add("doFinally " + any.toString()))
+		                          ::subscribe)
+		          .withMessage("java.lang.IllegalStateException: boom");
+
+		Assertions.assertThat(events)
+		          .as("whithoutDoOnError")
+		          .containsExactly("doFinally onError");
+	}
+
+	private Boolean throwError(Boolean x) {
+		throw new IllegalStateException("boom");
+	}
 }

